### PR TITLE
refactor: Separate database integration from `trie` into its own crate

### LINKED ISSUE #8514: Extract reth-trie-types from reth-trie
### Describe the feature

reth-trie mixes db implementations with type definitions, which is problematic if we want to use these types in evm for different targets (risc, maybe even no_std).

hence we should extract the core types to a standalone crate, primarily the TrieUpdates type

currently this type has functions that make use of db:

https://github.com/paradigmxyz/reth/blob/f43ccc760b150854654e7364892f6ec753c17eeb/crates/trie/trie/src/updates.rs#L132-L132

or 

https://github.com/paradigmxyz/reth/blob/f43ccc760b150854654e7364892f6ec753c17eeb/crates/trie/trie/src/updates.rs#L114-L119

The triewalker functions could be inverted and implemented on triewalker instead,
for the flush function we could use a tmp wrapper type instead

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by RomanHodulak**
Closes #8514

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
I'm unequipped to review these code changes and there are quite a few @rkrasiuk 

my original idea was to extract the types instead (especially trieupdates) and then work towards making them not dependent on the db deps.

but this approach also works, but I'm not sure about these changes, because using wrapper types to implement a trait is not ideal.

I assume it would be more useful to have
trie-types and trie
wdyt @rkrasiuk

**Comment by mattsse** [comment]
perhaps we kick this off by extracting these: 

https://github.com/paradigmxyz/reth/tree/01fd9b3c76705e582cea17774f269fe4a78d8af3/crates/primitives/src/trie

to trie-types instead and then try to move more stuff from reth-trie there?

File: crates/trie/trie-db/src/updates.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+};
+use std::collections::{hash_map::IntoIter, HashMap, HashSet};
+use reth_trie::updates::{TrieKey, TrieOp};
+use reth_trie::walker::TrieWalker;
+
+/// The aggregation of trie updates.
+#[derive(Debug, Default, Clone, PartialEq, Eq, Deref)]
+pub struct TrieUpdates {
+    trie_operations: HashMap<TrieKey, TrieOp>,
+}
```
[the remark this answers is not in the document]
Comment (RomanHodulak) [reply crates/trie/trie-db/src/updates.rs]: Addressed in 9c9fd19

File: crates/trie/trie-db/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,62 @@
+[package]
+name = "reth-trie-db"
```
[the remark this answers is not in the document]
Comment (RomanHodulak) [reply crates/trie/trie-db/Cargo.toml:2]: Thanks, I think it makes sense to have that suffix in a crate that's supposed to mean "DB integration of trie".

**Comment by RomanHodulak** [comment]
Hey @mattsse, thank you for reviewing my PR so swiftly.

So far I've been thinking that crates:
 * `reth-trie`  - contains types and behavior of the Trie domain, but does not depend on a database, the concept of database does not really exist there. Should I rename it to `reth-trie-types`?
 * `reth-trie-db` - is a projection of the `reth-trie` into the database, but does not define any concepts of the Trie domain. It implements the traits defined by the Trie domain using `reth_database`. There might be other implementations than the `reth_database`. that's why the `reth-trie-db` shouldn't contain domain behavior or types, otherwise that would lead to duplication of knowledge.

I haven't looked into the `primitives` crate yet, so that's where I'll go next. It does make sense to pull the `trie` module out of that crate and put it into the `reth-trie` crate.

**Comment by RomanHodulak** [comment]
Superseded by #9282

**Pull request closed without merging**